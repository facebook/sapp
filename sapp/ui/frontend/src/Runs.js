/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow
 */

import React, {useState} from 'react';
import {useQuery, useMutation, gql} from '@apollo/client';
import {Breadcrumb, Card, Col, Modal, Row, Typography} from 'antd';
import {DeleteOutlined, LoadingOutlined, SyncOutlined} from '@ant-design/icons';
import {clearFilter} from './Filter';

const {Text, Link} = Typography;

type RunDescription = $ReadOnly<{
  run_id: number,
  date: string,
}>;

function Run(props: $ReadOnly<{run: RunDescription}>): React$Node {
  const gutter = [8, 8];
  const [showRun, setShowRun] = useState(true);

  const Label = (props: $ReadOnly<{children: React$Node}>): React$Node => {
    return (
      <Col span={4} style={{textAlign: 'right'}}>
        <Text type="secondary">{props.children}</Text>
      </Col>
    );
  };
  const Item = (props: $ReadOnly<{children: React$Node}>): React$Node => {
    return (
      <Col span={20}>
        <Text type="secondary">{props.children}</Text>
      </Col>
    );
  };

  const deleteRunMutation = gql`
    mutation DeleteRun($id: ID!) {
      delete_run(input: {id: $id}) {
        clientMutationId
      }
    }
  `;

  const [deleteRun, {error: deleteError}] = useMutation(
    deleteRunMutation,
    {
      onCompleted() {
        setShowRun(false)
      }
    }
  );

  const onDelete = (): void => {
    deleteRun({variables: {id: props.run.run_id}});
  };

  if(deleteError) {
    Modal.error({title: 'Unable to delete run ', content: deleteError.toString()});
  }

  const contents = (
    <Col span={8}>
      <Card
        size="small"
        title={
          <>
            <SyncOutlined style={{marginRight: '.5em'}} />
            Run {props.run.run_id}
          </>
        }
        extra={
          <Link
            onClick={() => clearFilter()}
            href={`/run/${props.run.run_id}`}>
            Issues
          </Link>
        }
        actions={[
          <DeleteOutlined onClick={onDelete}/>
        ]}>
        <Row gutter={gutter}>
          <Label>Date</Label>
          <Item>
            <Text code>{props.run.date}</Text>
          </Item>
        </Row>
      </Card>
      <br />
    </Col>
  )

  return (
    <>
      { showRun ? contents : null }
    </>
  );
}

export const RunsQuery = gql`
  query Run {
    runs {
      edges {
        node {
          run_id
          date
        }
      }
    }
  }
`;

export default function Runs(props: $ReadOnly<{}>): React$Node {
  const {loading, error, data} = useQuery(RunsQuery);

  if (error) {
    Modal.error({title: 'Unable to load run data', content: error.toString()});
    return null;
  }

  var content = null;
  if (loading) {
    content = (
      <Col span={8}>
        <Card>
          <div style={{height: '12em', textAlign: 'center', paddingTop: '5em'}}>
            <Text type="secondary">
              <LoadingOutlined />
              <br />
              Loading runs...
            </Text>
          </div>
        </Card>
      </Col>
    );
  }

  if (data) {
    content = data.runs.edges.map(edge => <Run run={edge.node} />);
  }

  return (
    <>
      <Breadcrumb style={{margin: '16px 0'}}>
        <Breadcrumb.Item>Runs</Breadcrumb.Item>
      </Breadcrumb>
      <Row gutter={16}>
        {content}
      </Row>
    </>
  );
}
