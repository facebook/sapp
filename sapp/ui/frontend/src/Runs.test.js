/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 * @flow
 */

import React from 'react';
import TestRenderer from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import Runs, {RunsQuery} from './Runs';

const {act} = TestRenderer;

test("Renders runs", async () => {
  // Mock window.matchMedia
  window.matchMedia = window.matchMedia || function() {
    return {
      matches: false,
      addListener: function() {},
      removeListener: function() {}
    };
  };

  const mock = {
    request: {
      query: RunsQuery,
    },
    result: {
      data: {
        runs: {
          edges: [{
            node: {
              run_id: "1",
              date: "Dummy date",
            },
          }],
        },
      },
    },
  };

  const component = TestRenderer.create(
    <MockedProvider mocks={[mock]}>
      <Runs/>
    </MockedProvider>
  );

  // Mocked provider hasn't populated the results yet. Use it to test loading
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  /**
   * Await a zero-millisecond timeout. This delays the checks until the next
   * "tick" of the event loop, which gives MockedProvider an opportunity to
   * populate the mocked result
   */
  await act(() => new Promise((resolve) => setTimeout(resolve, 0)));
  tree = component.toJSON();
  expect(tree).toMatchSnapshot()
});
