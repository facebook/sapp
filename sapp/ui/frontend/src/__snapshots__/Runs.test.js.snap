// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders runs 1`] = `
Array [
  <div
    className="ant-breadcrumb"
    style={
      Object {
        "margin": "16px 0",
      }
    }
  >
    <span>
      <span
        className="ant-breadcrumb-link"
      >
        Runs
      </span>
      <span
        className="ant-breadcrumb-separator"
      >
        /
      </span>
    </span>
  </div>,
  <div
    className="ant-row"
    style={
      Object {
        "marginLeft": -8,
        "marginRight": -8,
      }
    }
  >
    <div
      className="ant-col ant-col-8"
      style={
        Object {
          "paddingLeft": 8,
          "paddingRight": 8,
        }
      }
    >
      <div
        className="ant-card ant-card-bordered"
      >
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            style={
              Object {
                "height": "12em",
                "paddingTop": "5em",
                "textAlign": "center",
              }
            }
          >
            <span
              className="ant-typography ant-typography-secondary"
              style={
                Object {
                  "WebkitLineClamp": null,
                }
              }
            >
              <span
                aria-label="loading"
                className="anticon anticon-loading"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className="anticon-spin"
                  data-icon="loading"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                  />
                </svg>
              </span>
              <br />
              Loading runs...
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Renders runs 2`] = `
Array [
  <div
    className="ant-breadcrumb"
    style={
      Object {
        "margin": "16px 0",
      }
    }
  >
    <span>
      <span
        className="ant-breadcrumb-link"
      >
        Runs
      </span>
      <span
        className="ant-breadcrumb-separator"
      >
        /
      </span>
    </span>
  </div>,
  <div
    className="ant-row"
    style={
      Object {
        "marginLeft": -8,
        "marginRight": -8,
      }
    }
  >
    <div
      className="ant-col ant-col-8"
      style={
        Object {
          "paddingLeft": 8,
          "paddingRight": 8,
        }
      }
    >
      <div
        className="ant-card ant-card-bordered ant-card-small"
      >
        <div
          className="ant-card-head"
          style={Object {}}
        >
          <div
            className="ant-card-head-wrapper"
          >
            <div
              className="ant-card-head-title"
            >
              <span
                aria-label="sync"
                className="anticon anticon-sync"
                role="img"
                style={
                  Object {
                    "marginRight": ".5em",
                  }
                }
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="sync"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M168 504.2c1-43.7 10-86.1 26.9-126 17.3-41 42.1-77.7 73.7-109.4S337 212.3 378 195c42.4-17.9 87.4-27 133.9-27s91.5 9.1 133.8 27A341.5 341.5 0 01755 268.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.7 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c0-6.7-7.7-10.5-12.9-6.3l-56.4 44.1C765.8 155.1 646.2 92 511.8 92 282.7 92 96.3 275.6 92 503.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8zm756 7.8h-60c-4.4 0-7.9 3.5-8 7.8-1 43.7-10 86.1-26.9 126-17.3 41-42.1 77.8-73.7 109.4A342.45 342.45 0 01512.1 856a342.24 342.24 0 01-243.2-100.8c-9.9-9.9-19.2-20.4-27.8-31.4l60.2-47a8 8 0 00-3-14.1l-175.7-43c-5-1.2-9.9 2.6-9.9 7.7l-.7 181c0 6.7 7.7 10.5 12.9 6.3l56.4-44.1C258.2 868.9 377.8 932 512.2 932c229.2 0 415.5-183.7 419.8-411.8a8 8 0 00-8-8.2z"
                  />
                </svg>
              </span>
              Run 
              1
            </div>
            <div
              className="ant-card-extra"
            >
              <a
                className="ant-typography"
                href="/run/1"
                onClick={[Function]}
                style={
                  Object {
                    "WebkitLineClamp": null,
                  }
                }
              >
                Issues
              </a>
            </div>
          </div>
        </div>
        <div
          className="ant-card-body"
          style={Object {}}
        >
          <div
            className="ant-row"
            style={
              Object {
                "marginBottom": 4,
                "marginLeft": -4,
                "marginRight": -4,
                "marginTop": -4,
              }
            }
          >
            <div
              className="ant-col ant-col-4"
              style={
                Object {
                  "paddingBottom": 4,
                  "paddingLeft": 4,
                  "paddingRight": 4,
                  "paddingTop": 4,
                  "textAlign": "right",
                }
              }
            >
              <span
                className="ant-typography ant-typography-secondary"
                style={
                  Object {
                    "WebkitLineClamp": null,
                  }
                }
              >
                Date
              </span>
            </div>
            <div
              className="ant-col ant-col-20"
              style={
                Object {
                  "paddingBottom": 4,
                  "paddingLeft": 4,
                  "paddingRight": 4,
                  "paddingTop": 4,
                }
              }
            >
              <span
                className="ant-typography ant-typography-secondary"
                style={
                  Object {
                    "WebkitLineClamp": null,
                  }
                }
              >
                <span
                  className="ant-typography"
                  style={
                    Object {
                      "WebkitLineClamp": null,
                    }
                  }
                >
                  <code>
                    Dummy date
                  </code>
                </span>
              </span>
            </div>
          </div>
        </div>
        <ul
          className="ant-card-actions"
        >
          <li
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <span>
              <span
                aria-label="delete"
                className="anticon anticon-delete"
                onClick={[Function]}
                role="img"
                tabIndex={-1}
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="delete"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M360 184h-8c4.4 0 8-3.6 8-8v8h304v-8c0 4.4 3.6 8 8 8h-8v72h72v-80c0-35.3-28.7-64-64-64H352c-35.3 0-64 28.7-64 64v80h72v-72zm504 72H160c-17.7 0-32 14.3-32 32v32c0 4.4 3.6 8 8 8h60.4l24.7 523c1.6 34.1 29.8 61 63.9 61h454c34.2 0 62.3-26.8 63.9-61l24.7-523H888c4.4 0 8-3.6 8-8v-32c0-17.7-14.3-32-32-32zM731.3 840H292.7l-24.2-512h487l-24.2 512z"
                  />
                </svg>
              </span>
            </span>
          </li>
        </ul>
      </div>
      <br />
    </div>
  </div>,
]
`;
